### PR TITLE
Provide IServiceProvider to FormStatus

### PR DESCRIPTION
**Pull request by mstv**
Fixes: preparation for git output / history control

## Proposed changes

- Ensure that `FormStatus` has a `GitUICommands` instance.
- Pass GitUICommands instance to FormStatus when running scripts by downcasting from IGitUICommands to GitUICommands
- Make GitUICommands instance mandatory for FormStatus & FormProcess

## Screenshots 

N/A

## Test methodology 

- existing tests

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/HelperDialogs/FormProcess.cs [hunk 1]
```diff
@@ -19,8 +19,15 @@ public partial class FormProcess : FormStatus
         public HandleOnExit? HandleOnExitCallback { get; set; }
         public readonly Dictionary<string, string> ProcessEnvVariables = new();
 
-        private FormProcess(GitUICommands? commands, ConsoleOutputControl? outputControl, ArgumentString arguments, string workingDirectory, string? input, bool useDialogSettings, string? process)
-            : base(commands, outputControl, useDialogSettings)
+        private FormProcess(IServiceProvider serviceProvider,
+                            GitUICommands? commands,
```
Comment (RussKie) [inline GitUI/HelperDialogs/FormProcess.cs:23]: Why are we doing this? `GitUICommands` is a service provider

Comment (mstv) [reply GitUI/HelperDialogs/FormProcess.cs:23]: `commands` is optional (can be `null`), `serviceProvider` is mandatory.

**Comment by RussKie** [comment]
Then we just need to make mandatory

**Comment by mstv** [comment]
> Then we just need to make mandatory

```
        protected GitModuleForm(GitUICommands? commands, bool enablePositionRestore)
            : base(enablePositionRestore)
        {
            if (commands is null && _uiCommands is null)
            {
                // In some cases we may want to initialise a form without a module,
                // e.g. a process dialog that executes a git command which is completely self contained.
            }
```

For making UICommands mandatory, we would need to switch to **I**GitUICommands instead of GitUICommands...
and YAGNI.

**Comment by RussKie** [comment]
Then we need to prioritise #11269 as it would likely unblock the use of the
interface.

I wonder what these "some cases"... Still, we can construct and pass an
instance of the commands.

**Comment by mstv** [comment]
> I wonder what these "some cases"...

Ask your self from 4 years ago ;-)
`// e.g. a process dialog that executes a git command which is completely self contained.`
This is true for all direct usages of FormStatus and FormProcess.

I do not see why FormStatus and all its descendants need to be a GitModuleForm. They have never used GitUICommands and should never do so except for IServiceProvider.
Even FormRemoteProcess uses its own property, only in order to access the subproperty `Module` for get & set settings but none of the many GitUICommands features nor UICommandsChanged, etc.

Though since an IGitUICommands - and thereby IServiceProvider - instance is already provided everywhere and since a refactoring of the grown inheritance chain will be even more difficult,
the anyway recommendable main topic of #11269 should be pursued.
FormStatus -> GitExtensionsDialog -> GitModuleForm -> GitExtensionsForm -> GitExtensionsFormBase (with TryGetUICommands)
I still feel uncomfortable with this overuse of IGitUICommands.

**Comment by RussKie** [comment]
I agree we should always pass an instance of IGitUICommands right now. In
the target state I don't want any UI elements to depend on that explicitly,
and instead get the required dependencies via the service provider.
I'm currently dealing with health issues, when I feel better I'll have a
look in more detail.

**Comment by mstv** [comment]
I have tried the naive approach to use the completed `IGitUICommands` instead of `GitUICommands` - as a quick first step ... I thought.
It will not work without completing `IGitModule`, too.
Raise event functions would need to be declared `public`: db5c1813e426146b28a23ae56d6f517a6ffbd479.
Several types would need to be moved to `GitUIPluginInterfaces`: 796cc110b29fe60821decf77a87433a814f365c4.

So I wait for #11269 and the follow-ups.

For the meantime, I am downcasting `IGitUICommands` to `GitUICommands` for calling `FormProcess.ShowDialog` in `RunScriptInternal`.

**Comment by RussKie** [comment]
> For the meantime, I am downcasting IGitUICommands to GitUICommands for
calling FormProcess.ShowDialog in RunScriptInternal.

Sounds good, just add a tracking comment.

**Pull request merged**